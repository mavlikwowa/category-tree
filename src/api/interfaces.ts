/** Свойства ответа метода getTree */
export interface GetTreeProps {
  /** Успешность ответа */
  success: boolean;
  /** Данные */
  data: CategoryProps[];
}

/** Свойство категории */
export interface CategoryProps {
  id: number;
  /** Отображаемое название */
  name: string;
  /** Мультипликатор цены */
  price_multiplier: number;
  /** id игры, в которой опубликована категория */
  game_id: number;
  /** timestmap создания */
  created_at: number;
  /** timestamp последнего обновления */
  updated_at: number;
  /** Порядок отображения */
  sort: number;
  /** Комиссия сервиса */
  commission: number;
  /** Показывать цену */
  show_price: number;
  /** Показывать количество */
  show_count: number;
  /** Показывать время */
  show_time: number;
  /** Можно продавать несколько раз */
  multiple_sale: number;
  /** Текущий статус: 0 - не видна, 1 - видна всем */
  status: number;
  /** ID родителя или null, если категория корневая */
  parent_id: number | null;
  /** Моментальная доставка? */
  is_instant_delivery: number;
  /** Можно создать несколько предложений в этой категории? */
  is_allow_multiple_offers: number | null;
  /** Есть ли вложенные категории: true/false */
  isFinal: boolean;
  /** Дочерние категории, которых может быть неограниченное количество */
  children?: CategoryProps[];
}
