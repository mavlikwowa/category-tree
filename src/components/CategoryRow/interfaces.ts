/** Интерфейсы */
import { CategoryProps } from '../../api/interfaces';
import { SetStateAction, Dispatch } from 'react';

export interface CategoryRowProps extends CategoryProps {
  /** Функция изменения флага */
  setShowChildren?: Dispatch<SetStateAction<boolean>>;
  /** Флаг отображения вложенных категорий */
  showChildren?: boolean;
}
