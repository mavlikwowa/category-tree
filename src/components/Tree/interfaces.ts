/** Интерфейсы */
import { CategoryProps } from '../../api/interfaces';
/** Actions */
import { changeTree } from '../../actionCreators/treeCreator';

/** Интерфейс dispatch'ей компонента */
export interface DispatchProps {
  changeTree: typeof changeTree;
}
/** Интерфейс state store компонента */
export interface StateProps {
  categories: CategoryProps[];
}
