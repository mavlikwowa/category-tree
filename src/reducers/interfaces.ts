import { CategoryProps } from '../api/interfaces';

export interface TreeReduxAction {
  type: TreeActionTypes;
  payload?: CategoryProps[];
}

/** Свойства Action Thunk */
export interface AppThunkAction<TAction> {
  (dispatch: <TAction>(action: TAction) => TAction, getState: () => any): void;
}

/** Список событий для treeReducer */
export enum TreeActionTypes {
  GET_TREE = 'GET_TREE',
}
