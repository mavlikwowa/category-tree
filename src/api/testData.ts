/** Интерфейсы */
import { GetTreeProps } from './interfaces';

/** Заглушка метода получения категорий */
export const TEST_TREE: GetTreeProps = {
  success: true,
  data: [
    {
      id: 6,
      name: 'Золото',
      price_multiplier: 1000,
      game_id: 3,
      created_at: 1555732503,
      updated_at: 1555732503,
      sort: 5,
      commission: 3,
      show_price: 1,
      show_count: 1,
      show_time: 1,
      multiple_sale: 1,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: null,
      isFinal: true,
    },
    {
      id: 7,
      name: 'Рейды',
      price_multiplier: 1,
      game_id: 3,
      created_at: 1555816711,
      updated_at: 1555816711,
      sort: 6,
      commission: 10,
      show_price: 0,
      show_count: 0,
      show_time: 0,
      multiple_sale: 0,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: null,
      isFinal: false,
      children: [
        {
          id: 8,
          name: "Дазар'Алор",
          price_multiplier: 1,
          game_id: 3,
          created_at: 1555817073,
          updated_at: 1555817073,
          sort: 1,
          commission: 10,
          show_price: 0,
          show_count: 0,
          show_time: 0,
          multiple_sale: 1,
          status: 1,
          parent_id: 7,
          is_instant_delivery: 0,
          is_allow_multiple_offers: null,
          isFinal: true,
        },
        {
          id: 9,
          name: 'Горнило штормов',
          price_multiplier: 1,
          game_id: 3,
          created_at: 1555919373,
          updated_at: 1555919373,
          sort: 3,
          commission: 10,
          show_price: 0,
          show_count: 1,
          show_time: 1,
          multiple_sale: 1,
          status: 1,
          parent_id: 7,
          is_instant_delivery: 0,
          is_allow_multiple_offers: null,
          isFinal: true,
        },
      ],
    },
    {
      id: 19,
      name: 'Серебро',
      price_multiplier: 1000,
      game_id: 3,
      created_at: 1559631163,
      updated_at: 1559631163,
      sort: 13,
      commission: 10,
      show_price: 1,
      show_count: 1,
      show_time: 1,
      multiple_sale: 1,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: 1,
      isFinal: true,
    },
    {
      id: 51,
      name: 'Серебро',
      price_multiplier: 1000,
      game_id: 3,
      created_at: 1573635179,
      updated_at: 1574063841,
      sort: 25,
      commission: 10,
      show_price: 1,
      show_count: 1,
      show_time: 1,
      multiple_sale: 1,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: null,
      isFinal: true,
    },
  ],
};
