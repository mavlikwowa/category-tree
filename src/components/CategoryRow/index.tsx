import React, { FC } from 'react';
import format from 'date-fns/format';
import { CategoryRowProps } from './interfaces';

/** Дочерний компонент "Строка с категорией" */
const CategoryRow: FC<CategoryRowProps> = ({
  name,
  price_multiplier,
  game_id,
  created_at,
  updated_at,
  commission,
  show_price,
  show_count,
  show_time,
  multiple_sale,
  status,
  is_instant_delivery,
  is_allow_multiple_offers,
  isFinal,
  showChildren,
  setShowChildren,
}: CategoryRowProps) => {
  /** Показывает вложенные категории */
  const handlerChildren = () => {
    if (setShowChildren && typeof showChildren === 'boolean') {
      setShowChildren(!showChildren);
    }
  };

  /** Формирует дату из timestamp*/
  const createDateFromTimestamp = (timestamp: number) => {
    return format(new Date(timestamp * 1000), 'dd/LL/yyyy HH:mm:SS');
  }

  return (
    <>
      <td>{name}</td>
      <td>{price_multiplier}</td>
      <td>{game_id}</td>
      <td>{createDateFromTimestamp(created_at)}</td>
      <td>{createDateFromTimestamp(updated_at)}</td>
      <td>{commission}</td>
      <td>{Boolean(show_price) ? 'Да' : 'Нет'}</td>
      <td>{Boolean(show_count) ? 'Да' : 'Нет'}</td>
      <td>{Boolean(show_time) ? 'Да' : 'Нет'}</td>
      <td>{Boolean(multiple_sale) ? 'Да' : 'Нет'}</td>
      <td>{Boolean(status) ? 'Виден' : 'Скрыт'}</td>
      <td>{Boolean(is_instant_delivery) ? 'Да' : 'Нет'}</td>
      <td>{Boolean(is_allow_multiple_offers) ? 'Можно' : 'Нельзя'}</td>
      {isFinal ? <td /> :
        <td
          onClick={handlerChildren}
          className={`${!isFinal && 'show_button'}`}
        >
          {!isFinal ? `${showChildren ? 'Скрыть' : 'Показать'} вложенные категории` : ''}
        </td>}
    </>
  );
};

export default CategoryRow;
