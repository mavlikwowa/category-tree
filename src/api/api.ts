/** Библиотеки */
import axios from 'axios';
/** Константы */
import { API_URL } from './config';
import { TEST_TREE } from './testData';
/** Интерфейсы */
import { CategoryProps } from './interfaces';

/**
 * Возвращает дерево категорий
 */
export const getTree = async (): Promise<CategoryProps[] | void> => {
  try {
    // const res = await axios.get(`${API_URL}/categories/tree`);
    // if (!res.data.success) throw 'request error';
    // return res.data.data;
    const res = await new Promise(resolve => setTimeout(() => { resolve(TEST_TREE.data) }, 1000));
    return res as CategoryProps[];
  } catch (e) {
    console.log(e);
  }
};

/**
 * Меняет позицию категории
 * @param categoryId - идентификатор категории
 * @param sort - индекс сортировки
 */
export const changeCategory = async (
  categoryId: number,
  sort: number,
): Promise<void> => {
  try {
    const res = await axios({
      method: 'post',
      url: `${API_URL}/categories/${categoryId}/sort/${sort}`,
      headers: {},
    })
    // if (!res.data.success) throw 'request error';
  } catch (e) {
    console.log(e);
  }
}
