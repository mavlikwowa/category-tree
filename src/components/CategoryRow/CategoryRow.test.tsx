/* Библиотеки */
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

/* Компоненты */
import CategoryRow from './index';
/* Интерфейсы */
import { CategoryRowProps } from './interfaces';
/* Тестовые данные */
import { TEST_TREE } from '../../api/testData';

/** Настройки */
const setUp = () => {
  /* mock для функции setShowChildren */
  const mockSetShowChildren = jest.fn();
  /* Тестовые данные */
  const props: CategoryRowProps = {
    ...TEST_TREE.data[0],
    setShowChildren: mockSetShowChildren,
    showChildren: true,
  };
  /* Компонент */
  const wrapper = shallow(
      <CategoryRow {...props} />
  );

  return { wrapper, props };
};

describe('Test component Link', () => {
  let wrapper: ShallowWrapper;
  let props: CategoryRowProps;

  beforeEach(() => {
    ({ wrapper, props } = setUp());
  });

  afterEach(() => {
    if (!wrapper) return;
    wrapper.unmount();
  });

  it('should render component', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
