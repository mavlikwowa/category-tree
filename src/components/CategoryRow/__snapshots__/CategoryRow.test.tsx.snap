// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component Link should render component 1`] = `
<Fragment>
  <td>
    Золото
  </td>
  <td>
    1000
  </td>
  <td>
    3
  </td>
  <td>
    20/04/2019 08:55:00
  </td>
  <td>
    20/04/2019 08:55:00
  </td>
  <td>
    3
  </td>
  <td>
    Да
  </td>
  <td>
    Да
  </td>
  <td>
    Да
  </td>
  <td>
    Да
  </td>
  <td>
    Виден
  </td>
  <td>
    Нет
  </td>
  <td>
    Нельзя
  </td>
  <td />
</Fragment>
`;
